[SESK-1] Scripted Installation Process for the Engage Online Learners Starter Kit

jleiva-gap opened the pull request:
- Update configuration.json to add new parameters.
- Add specific modules to install: BulLoadClient, Admin, AMT, DB, Swagger, WebAPI.
- Update scripts to use database from configuration.json.
- Remove hardcoded connectionstrings and application names.
- Update to version 5.3.
- Update to connect to remote databases (it doesn't work to install Databases because the installer cannot restore a database from a remote machine).
- Update scripts to use config file as a common parameter (from ps1 files to psm1)

stephenfuqua commented on scripts/Install-EdFiTechnologySuite.ps1:
```diff
@@ -59,97 +59,181 @@
-    # NuGet Feed for Ed-Fi packages
-    [string]
-    $EdFiNuGetFeed = "https://pkgs.dev.azure.com/ed-fi-alliance/Ed-Fi-Alliance-OSS/_packaging/EdFi%40Release/nuget/v3/index.json"
+    [string] $configPath = "$PSScriptRoot\configuration.json"
```
@jleiva-gap why would you remove all of these parameters?  That was very carefully designed.

stephenfuqua replied:
OK, I suppose I see the answer: because it is easier to pass in the configuration file. My initial reaction is not to do this, as that is a non-standard way of working with PowerShell functions. Someone trying to run `get-help ./Install-EdFiTechnologySuite.ps1`, for example, will just see that they need to supply a config file. But they won't know what goes into that config file, whereas with the arguments, they'll get a complete description of the arguments they can provide.

stephenfuqua replied:
@jleiva-gap keeping the config file is fine at this point, since this is now your team's code. However, one suggestion: validate that the config file actually has the expected parameters in it before trying to use those parameters.

jleiva-gap replied:
I updated the code to not take the config file by default, it was replaced by default values in the parameters. I removed all the references to the configuration.json file, we can use it, but it is an option.

stephenfuqua commented on scripts/modules/BulkLoadClient.psm1:
```diff
@@ -0,0 +1,52 @@
+# See the LICENSE and NOTICES files in the project root for more information.
+
+#Requires -Version 5
+#Requires -RunAsAdministrator
```
Does this really need to run as administrator?

Is this function just split out from an existing module, or is it entirely new?

jleiva-gap replied:
I just moved that function from Install-EdFiPlatform, I think it doesn't need to run as administrator because the current version doesn't have that line.

stephenfuqua commented on scripts/modules/nuget-helper.psm1:
```diff
@@ -0,0 +1,150 @@
+# See the LICENSE and NOTICES files in the project root for more information.
+#Requires -version 5
+param(
+    [parameter(Position=0,Mandatory=$true)][Hashtable]$configuration
+)
+$packageSource = If ($configuration.EdFiNuGetFeed) {"$($configuration.EdFiNuGetFeed)"} Else {"https://pkgs.dev.azure.com/ed-fi-alliance/Ed-Fi-Alliance-OSS/_packaging/EdFi%40Release/nuget/v3/index.json"}
```
This is a bit hard to read. Perhaps improve like...

```suggestion
    [parameter(Position=0,Mandatory=$true)]
    [Hashtable]
    $configuration
)
$default = "https://pkgs.dev.azure.com/ed-fi-alliance/Ed-Fi-Alliance-OSS/_packaging/EdFi%40Release/nuget/v3/index.json"
$override = $configuration.EdFiNuGetFeed

$packageSource = If ($override ) { $override } Else { $default }
```

Of course, that's only if keeping the configuration file 😉 .

stephenfuqua commented on scripts/modules/nuget-helper.psm1:
```diff
@@ -0,0 +1,150 @@
+    return "$ToolsPath/$PackageName.$PackageVersion"
+}
+
+Export-ModuleMember *
```
Please add documentation strings for exported members.

DavidJGapCR commented:
Noticed that an extra database is installed. Is that ok?
![image](https://user-images.githubusercontent.com/16614428/157348202-4fdecc0e-00a6-478a-877d-d0480801e844.png)

DavidJGapCR commented:
It looks like wasn't installed successfully.
![image](https://user-images.githubusercontent.com/16614428/157349535-2e01a526-ef1c-4d3d-a726-c20d3b816d5b.png)

This is the packer file:
[packer.log](https://github.com/Ed-Fi-Alliance-OSS/Engage-Online-Learners-Starter-Kit/files/8210697/packer.log)

stephenfuqua commented:
> Noticed that an extra database is installed. Is that ok? 

That database should only exist when installing the Sandbox Admin, so no, it is preferable that it not exist.

jleiva-gap commented:
> It looks like wasn't installed successfully. ![image](https://user-images.githubusercontent.com/16614428/157349535-2e01a526-ef1c-4d3d-a726-c20d3b816d5b.png)
> 
> This is the packer log file: [packer.log](https://github.com/Ed-Fi-Alliance-OSS/Engage-Online-Learners-Starter-Kit/files/8210697/packer.log)

I fixed that. Also I fixed the databases installation, now it adds just three databases.

DavidJGapCR commented:
Ran the scripts again and the 2 errors I reported have been fixed. It looks good to me. I'll go ahead and approve it.
